# ONT_Methylation
capture scripts and information for methylation study using ONT modified bases basecall

The main script is process_gene_features.py.
It processes a number of input files, described below, to generate files with C-methylation (Cm) information in relation to chromosome features and CpG islands.

Script usage: ./process_gene_features.py chromosome<X>_features
This script will process 5 input files to extract the Cm probability for the features in relation to the CpG islands  identified in the chromosome <X>.

The 5 input files are:
  - chromosome<X>_features: gene feature file derived from the Ensembl gff for fAstCal1.2
    The format is <feature> <name> <strand> <start> <end>:
    gene ID=gene:ENSACLG00000000002;Name=lto1;biotype=protein_coding;description=LTO1 maturation factor of ABCE1 [Source:ZFIN%3BAcc:ZDB-GENE-030729-33];gene_id=ENSACLG00000000002;logic_name=ensembl;version=1 - 19687 21631
    mRNA ID=transcript:ENSACLT00000000002;Parent=gene:ENSACLG00000000002;Name=lto1-201;biotype=protein_coding;transcript_id=ENSACLT00000000002;version=1 - 19687 21631
  - CpG_chr<X>: CpG island file derived from the output of CpGProd_linux for each chromosome. The CpG islands for each chromosome were named CpG<X> where X is an incremental integer starting at 1
    The format is <start> <end> <name>:
    1 912 CpG1
    1429 2152 CpG2
  - fastq_mapping_sorted.bam: binary version of the sam file generated by mapping of the fastq reads to the genome with minimap2
  - two pre-processed files: one containing the position of CG dinucleotides in each reads of the fastq files obtained after basecalling and a second one containing the corresponding Cm probability at these positions. They were obtained by running preprocess_modif_tables.py. Both files are in fasta like format:
   + Preprocessed_Table_pos:
     >4926de3b-22bf-45fe-8234-7f58eb002bc5
     37, 196, 243, 389, 714
     >591e2997-d484-4ba2-a129-d027f4d18e6b
     506, 510, 530, 549, 552, 569, 606, 629, 679, 701, ...
   + Preprocessed_Table_mod:
    >001540a0-54e5-41fb-a00a-08db26ff5e62
    254, 252, 205, 215, 254, 242, 252, 252, 253, 228, 254, 254, 250, 234, 210, 85, 253, 251, 167, 153,....
The path to the folders to access these files are provided into the contig file: process.config

The time for each step will be captured into the process_chr<X>.log file.

It will produce a number of files ouput:
   - chr<X>_CpG_<feature>_results.tsv lists the feature name, length then Cm probability, number of CG dinucleotide and cumulative length of regions within and outside the CpG islands
     feature length  Cm_noCpG        nberCG_noCpG    length_noCpG    Cm_inCpG        nberCG_inCpG    length_inCpG
     ENSACLG00000020698      864     46963.6 447.1   32613.2 0.0     0.0     0.0 ...
   - chr<X>_CpG_<feature>_summary.tsv lists the feature name, length then Cm probability for the feature and regions within and outside the CpG islands normalised by the feature length
     feature length  Cm_per_feature_length   Cm_noCpG_per_feature_length     Cm_inCpG_per_feature_length
     ENSACLG00000020698      84     622.1   1244.2  0 ...
   - chr<X>_feature_results.tsv summarize the results for each type of feature across all the chromosome
     feature total_number    mean_length     mean_prob       mean_prob/length        noCpG_mean_prob noCpG_mean_length       noCpG_mean_CGcount      features_with_CpG       inCpG_mean_length       inCpG_mean_CGcount
     PROMOTER        2       254.0   27921.8 219.9   27780.6 16426.9 240.7   141.1   141.0   3.0
     5prime_UTR      2       3114.0  35300.8 22.7    34957.3 14507.7 268.0   343.5   152.2   3.8 ...
   - Cm_chr<X>.bed list the coodinates of the features and their Cm probabilities
     chr1    19687   21573   4341.4
     chr1    21631   78636   118097.4 ...
   - Cm_inCpG_chr<X>.bed list the coodinates of the features and the Cm probabilities associated with the CpG islands regions
     chr1    19687   21573   3489.5
     chr1    21631   78636   99099.0
The path to the folders to store these files are also provided into the contig file: process.config
